core/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -0,0 +1,394 @@
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ */
+public class Setting<T> extends ToXContentToBytes {
+    private final String key;
+    private final String description;
+    private final Function<Settings, String> defaultValue;
+    private final Function<String, T> parser;
+    private final boolean dynamic;
+    private final Scope scope;
+
+    public Setting(String key, String description, Function<Settings, String> defaultValue, Function<String, T> parser, boolean dynamic, Scope scope) {
+        this.key = key;
+        this.description = description;
+        this.defaultValue = defaultValue;
+        this.parser = parser;
+        this.dynamic = dynamic;
+        this.scope = scope;
+    }
+
+    /**
+     * Returns the settings key or a prefix if this setting is a group setting
+     * @see #isGroupSetting()
+     */
+    public String getKey() {
+        return key;
+    }
+
+    /**
+     * Returns a human readable description of this setting
+     */
+    public String getDescription() {
+        return description;
+    }
+
+    /**
+     * Returns <code>true</code> iff this setting is dynamically updateable, otherwise <code>false</code>
+     */
+    public boolean isDynamic() {
+        return dynamic;
+    }
+
+    /**
+     * Returns the settings scope
+     */
+    public Scope getScope() {
+        return scope;
+    }
+
+    /**
+     * Returns <code>true</code> iff this setting is a group setting. Group settings represent a set of settings
+     * rather than a single value. The key, see {@link #getKey()}, in contrast to non-group settings is a prefix like <tt>cluster.store.</tt>
+     * that matches all settings with this prefix.
+     */
+    public boolean isGroupSetting() {
+        return false;
+    }
+
+    /**
+     * Returns the default values string representation for this setting.
+     * @param settings a settings object for settings that has a default value depending on another setting if available
+     */
+    public String getDefault(Settings settings) {
+        return defaultValue.apply(settings);
+    }
+
+    /**
+     * Returns <code>true</code> iff this setting is present in the given settings object. Otherwise <code>false</code>
+     */
+    public boolean exists(Settings settings) {
+        return settings.get(key) != null;
+    }
+
+    /**
+     * Returns the settings value. If the setting is not present in the given settings object the default value is returned
+     * instead.
+     */
+    public T get(Settings settings) {
+        String value = getRaw(settings);
+        try {
+            return parser.apply(value);
+        } catch (ElasticsearchParseException ex) {
+            throw ex;
+        } catch (Throwable t) {
+            throw new IllegalArgumentException("Failed to parse value [" + value + "] for setting [" + getKey() + "]", t);
```

Comment:
I don't think it makes a big difference really. I figure if the thing already throws an ElasticsearchParseException then why not throw more of them? But I don't think its a huge deal.

Also, Throwable will catch weird stuff like oom, right? So just catch Exception?